promql/engine.go

```diff
@@ -1371,6 +1525,9 @@ func (ev *evaluator) vectorSelector(node *parser.VectorSelector, ts int64) (Vect
 
 // vectorSelectorSingle evaluates a instant vector for the iterator of one time series.
 func (ev *evaluator) vectorSelectorSingle(it *storage.BufferedSeriesIterator, node *parser.VectorSelector, ts int64) (int64, float64, bool) {
+	if node.Timestamp > 0 {
```

Comment:
Why is this code here? Shouldn't this already have this value?

Reply:
It is here because it's simpler to handle it at this level. Or is it better to handle it at a higher level? We are just passing the evaluation time to `vectorSelectorSingle` right now.

Reply:
My point is that you have already handled this code at a higher level, the evaluation time should already be correct when you created a new evaluator.

Reply:
> the evaluation time should already be correct when you created a new evaluator.

Nope, the evaluation time is always w.r.t the actual evaluation times. Since the step invariant node is at a higher level, if a vector selector had `@` inside a subquery with `@`, we need it to be handled it here. So altogether, `@` will be handled at lower instead and not when we create a new evaluator.

Reply:
Hmm, that may be the case.

I'm overall finding this code very hard to review. There's no comments explaining how all the evaluation flow is meant to tie together, I have to infer your entire design from the code alone, and then try to figure out if it's correct which is quite challenging.